[pt1][quant] Add dynamic quantized Linear op in PyTorch

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23139 [pt1][quant] Unit Test for the Dynamic Quantized Linear operator
* #24231 [WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path
* #24299 [pt1][quant] Remove the activation observer for default_qconfig
* #24232 [pt1][quant] Make the default qconfig_dict
* #23330 [pt1][quant] Fix the dimension mismatch issues when running the BERT model
* #23212 [pt1][quant] Add the type matching rule for qconfig_dict
* #23128 [pt1][quant] Dynamic Quantized Linear module
* **#23464 [pt1][quant] Add dynamic quantized Linear op in PyTorch**

As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
[pt1][quant] Add dynamic quantized Linear op in PyTorch

As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous Diff D16381552 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -273,6 +273,50 @@ def test_cat(self, X, num, axis, relu):
+        use_relu=st.booleans(),
+    )
+    def test_qlinear(self, use_bias, use_relu):
+        batch_size = 1
```
Please randomize over batch size, input and output channels. We should check that:
 ```python
# Observe X_fp32 and determine scale and zero-point, this should match internals of dynamic linear.
 X_q = torch.quantize_linear(X_fp32 , scale , zero_point)
Y_fp32_ref = F.linear(X_q.dequantize(), W_q.dequantize(), b_fp32)
Y_fp32 = qlinear_dynamic(X_q.dequantize(), W_prepack,b_fp32)
```

jianyuh replied:
Thanks @raghuramank100 . As discussed in https://github.com/pytorch/pytorch/pull/23104, we created a task for the unit test for the exact match between static and dynamic quantization. Currently, if we use a randomized tensor as you suggested, we cannot get an exact match with `assert_equal`.

raghuramank100 reviewed (changes requested):
Please address comments, thanks

jianyuh commented:
> Please address comments, thanks

Currently we use a special example (the same as https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L11177-L11198). Maybe let's postpone the general way of unit test to unblock other dynamic quantization operator implementations. This PR is exactly the same as https://github.com/pytorch/pytorch/pull/23104, except it fixes a previous OSS build error (`quantize_linear` expects the `scale` to be float, and the `zero_point` to be int.)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh commented:
@raghuramank100 Ping for stamp to unblock: this PR is the same as https://github.com/pytorch/pytorch/pull/23104.

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh commented:
Request for review @raghuramank100

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -311,6 +311,113 @@ def test_cat(self, X, num, axis, relu):
+
+        # Observe X_fp32 and determine X_scale and X_zero_point, this should match
+        # internals of dynamic linear.
+        X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8)
```
Can you add an additional test to check that _calculate_dynamic_qparams works as expected for one sided distributions: please check for both distributions that are positive and distributions that are negative.

jianyuh replied:
Sorry I might misunderstand this: we use `_calculate_dynamic_qparams` to calculate the quantization parameters for both activation (`torch.quint8`) and weight (`torch.qint8`). What do "both distributions" refer to here?

As mentioned above, `_calculate_dynamic_qparams` is from https://github.com/pytorch/pytorch/blob/master/torch/quantization/observer.py#L39-L64.

jianyuh replied:
For `_calculate_dynamic_qparams`,  we will include 0 in the range. Is this "both distributions" you referred to?

raghuramank100 replied:
Yes, that resolves it.

jianyuh replied:
Should I add a test for `_calculate_dynamic_qparams`? The `_calculate_dynamic_qparams` is a test utility function, not a PyTorch API. Let me know if you want me to add a separate test for the test utility function itself.

raghuramank100 replied:
No, there is no need for a test for this function

raghuramank100 commented on test/common_quantized.py:
```diff
@@ -34,3 +36,24 @@ def _requantize(x, multiplier, zero_point, qmin=0, qmax=255, qtype=np.uint8):
+        scale = 1.0
+        zero_point = 0
+    else:
+        scale = (max_val - min_val) / n_levels
```
Should be:
```
max_val = torch.max(max_val, torch.tensor([0.0])
min_val = torch.min(min_val,torch.tensor([0.0])
```
 I dont see the code for fbGemm::choosequantizationparams or findminmax, but we need to ensure that this logic is implemented there too.

jianyuh replied:
Why do we need to compare the `max_val` and `min_val` with 0.0?

I followed the same logic in the observer for calculating the quantization parameters:
https://github.com/pytorch/pytorch/blob/master/torch/quantization/observer.py#L39-L64

jianyuh replied:
Oh, I know why we need to compare with 0 here: We extend the [min, max] interval to ensure that it contains 0. Otherwise, we would not meet the requirement that 0 be an exactly representable value.

The logic for fbgemm::choosequantizationparams is here:
https://github.com/pytorch/FBGEMM/blob/master/src/QuantUtils.cc#L43-L47

jianyuh replied:
Updated the code.

raghuramank100 replied:
The observer code has an error that I am fixing, wanted to make sure it doesnt propagate.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -311,6 +311,113 @@ def test_cat(self, X, num, axis, relu):
+        # print("Y_fp32:")
+        # print(Y_fp32)
+        # print("Y_fp32_ref:")
+        # print(Y_fp32_ref)
```
NIT: Please remove comments

jianyuh replied:
Will remove.

jianyuh replied:
Removed.

raghuramank100 reviewed (changes requested):
Please see comments, thanks

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -311,6 +311,100 @@ def test_cat(self, X, num, axis, relu):
+            _dequantize(b_q0, X_scale * W_scale, 0)
+        ).to(dtype=torch.float) if use_bias else None
+
+        W_scale, W_zp = _calculate_dynamic_qparams(W_fp32, torch.qint8)
```
These are known beforehand, why calculate them here?

jianyuh replied:
This is for to mimic the process of calculating the dynamaic quantization parameters. For this test, the input should be `X_fp32`, `W_fp32`, `b_fp32` (after the special construction). You can note that we have a `avoid_vpmaddubsw_overflow_linear` function in Line 366, which avoids the test cases with overflowing the vpmaddubsw int16_t ranges. The `W_scale` and `W_zp` might change with the update of `W_q0` in that function.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -311,6 +311,100 @@ def test_cat(self, X, num, axis, relu):
+
+        # Observe X_fp32 and determine X_scale and X_zero_point, this should match
+        # internals of dynamic linear.
+        X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8)
```
Suggest doing the following to verify that dynamic quantization of activations is done correctly
```python
#x_min and x_max are random float values that we should sweep over, including cases where x_min and x_max are both positive and negative.
x = np.random.rand(batch_size, input_channels)*(x_max-x_min)+x_min
# w_max and w_min can be fixed for this test as w is quantized to 8 bits outside of the dynamic linear op.
w = np.random.rand(output_channels, input_channels)*(w_max-w_min)+w_min

X_scale, X_zp =  _calculate_dynamic_qparams(x,torch.quint8)
x_fp32 = torch.quantize_linear(x, X_scale, X_zp, torch.quint8).dequantize()
W_q = torch.quantize_linear(w,.W_scale, W_zp, torch.qint8)
W_prepack = qlinear_prepack(W_q)
Y_fp32 = qlinear_dynamic(X_fp32, W_prepack, b_fp32)
Y_fp32_ref = F.linear(X_fp32, W_q.dequantize(), b_fp32)
```

jianyuh replied:
We cannot handle the random generated `X` and `W`. As mentioned before, we have to call `avoid_vpmaddubsw_overflow_linear` to specially handle that. If we don't call that, we will have the overflow issues.

I think the logic here is the same as the current code. Where is the code piece to verify the dynamic quantization of activation (my understanding is to check 0 is in the range of `x_min` and `x_max`. Correct me if I am wrong)?

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)

jianyuh committed:
Update on "[pt1][quant] Add dynamic quantized Linear op in PyTorch"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for torch.fbgemm_linear_int8_weight (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

The previous PR  https://github.com/pytorch/pytorch/pull/23104 is reverted because `quantize_linear` expects the scale to be `float`, and the zero_point to be `int`.

Differential Revision: [D16527741](https://our.internmc.facebook.com/intern/diff/D16527741/)
